Bottom: bba1cc5e7b2059569f2138ef2b5574c6d183a012
Top:    ac6413e141fb717a618cb12aa2415c09a65291c0
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-02-08 21:17:12 +0000

Manually count the vertices in a contour, report if the polygon code is lying



---

diff --git a/src/hid/common/hidgl.c b/src/hid/common/hidgl.c
index 7106410..0533240 100644
--- a/src/hid/common/hidgl.c
+++ b/src/hid/common/hidgl.c
@@ -648,8 +648,22 @@ hidgl_fill_pcb_polygon (PolygonType *poly)
   /* Walk the polygon structure, counting vertices */
   piece = poly->Clipped;
   do {
-    for (contour = piece->contours; contour != NULL; contour = contour->next)
+    for (contour = piece->contours; contour != NULL; contour = contour->next) {
+      int contour_count_manual = 0;
+      vnode = &contour->head;
+      do {
+        contour_count_manual++;
+      } while ((vnode = vnode->next) != &contour->head);
+
+      if (contour_count_manual != contour->Count) {
+        fprintf (stderr, "CRAPPY POLYGON ROUTINES MISCOUNTED NODES, FIXING\n");
+        fprintf (stderr, "Manual count was %i, contour claimed %i\n",
+                 contour_count_manual, contour->Count);
+        contour->Count = contour_count_manual;
+      }
+
       vertex_count += contour->Count;
+    }
   } while ((piece = piece->f) != poly->Clipped);
 
   vertices = malloc (sizeof(GLdouble) * vertex_count * 3);
